hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -321,14 +319,25 @@ public void run(boolean cancelled) throws IOException {
         if (cancelled) {
           setCancelled(call);
         } else {
+          Channel channel = channelRef.get();
           if (channel == null) {
-            connect();
+            try {
+              Channel newChannel = connect();
+              if (!channelRef.compareAndSet(null, newChannel)) {
+                newChannel.close();
+              }
+              channel = channelRef.get();
```

Comment:
What prevents `channel` from becoming `null` at this point again? An `AtomicBoolean` for connecting would be better in my opinion.

Reply:
Shall we just add an extra null check and skip sending the call if channel has already been closed?
Would you please give me more guidance on the AtomicBoolean idea?